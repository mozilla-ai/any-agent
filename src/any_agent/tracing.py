import os
import json
from datetime import datetime

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from opentelemetry.sdk.trace.export import SpanExporter


class JsonFileSpanExporter(SpanExporter):
    def __init__(self, file_name: str):
        self.file_name = file_name
        # Initialize with an empty array if file doesn't exist
        if not os.path.exists(self.file_name):
            with open(self.file_name, "w") as f:
                json.dump([], f)

    def export(self, spans) -> None:
        # Read existing spans
        try:
            with open(self.file_name, "r") as f:
                all_spans = json.load(f)
        except (json.JSONDecodeError, FileNotFoundError):
            all_spans = []

        # Add new spans
        for span in spans:
            try:
                # Try to parse the span data from to_json() if it returns a string
                span_data = json.loads(span.to_json())
            except (json.JSONDecodeError, TypeError, AttributeError):
                # If span.to_json() doesn't return valid JSON string
                span_data = {"error": "Could not serialize span", "span_str": str(span)}

            all_spans.append(span_data)

        # Write all spans back to the file as a proper JSON array
        with open(self.file_name, "w") as f:
            json.dump(all_spans, f, indent=2)

    def shutdown(self):
        pass


def get_tracer_provider(
    project_name: str, output_dir: str = "telemetry_output"
) -> tuple[TracerProvider, str | None]:
    """
    Create a tracer_provider based on the selected mode.

    Args:
        project_name: Name of the project for tracing
        output_dir: The directory where the telemetry output will be stored.
            Only used if `json_tracer=True`.
            Defaults to "telemetry_output".

    Returns:
        tracer_provider: The configured tracer provider
        file_name: The name of the JSON file where telemetry will be stored
    """
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    timestamp = datetime.now().strftime("%Y-%m-%d-%H-%M-%S")

    tracer_provider = TracerProvider()
    trace.set_tracer_provider(tracer_provider)

    file_name = f"{output_dir}/{project_name}-{timestamp}.json"
    json_file_exporter = JsonFileSpanExporter(file_name=file_name)
    span_processor = SimpleSpanProcessor(json_file_exporter)
    tracer_provider.add_span_processor(span_processor)

    return tracer_provider, file_name


def setup_tracing(tracer_provider: TracerProvider, agent_framework: str) -> None:
    """Setup tracing for `agent_framework` by instrumenting `trace_provider`.

    Args:
        tracer_provider (TracerProvider): The configured tracer provider from
            [get_tracer_provider][surf_spot_finder.tracing.get_tracer_provider].
        agent_framework (str): The type of agent being used.
            Must be one of the supported types in [RUNNERS][surf_spot_finder.agents.RUNNERS].
    """
    if "openai" in agent_framework:
        from openinference.instrumentation.openai_agents import OpenAIAgentsInstrumentor

        OpenAIAgentsInstrumentor().instrument(tracer_provider=tracer_provider)
    elif agent_framework == "smolagents":
        from openinference.instrumentation.smolagents import SmolagentsInstrumentor

        SmolagentsInstrumentor().instrument(tracer_provider=tracer_provider)
    elif agent_framework == "langchain":
        from openinference.instrumentation.langchain import LangChainInstrumentor

        LangChainInstrumentor().instrument(tracer_provider=tracer_provider)
    else:
        raise NotImplementedError(f"{agent_framework} tracing is not supported.")
