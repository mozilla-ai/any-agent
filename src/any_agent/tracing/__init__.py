from .cost import TokenUseAndCost, TotalTokenUseAndCost, extract_token_use_and_cost
from .processors.base import TracingProcessor

__all__ = [
    "TokenUseAndCost",
    "TotalTokenUseAndCost",
    "TracingProcessor",
    "extract_token_use_and_cost",
]
