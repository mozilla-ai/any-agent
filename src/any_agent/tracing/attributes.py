"""Constants exported for convenience to access span attributes.

Trying to follow OpenTelemetry's [Semantic Conventions for Generative AI](https://opentelemetry.io/docs/specs/semconv/gen-ai/).
"""

INPUT_MESSAGES = "gen_ai.input.messages"
"""System prompt and user input."""

MODEL_ID = "gen_ai.request.model"
"""Underlying (LLM) model used by the agent. For example: `gpt-4.1-mini`"""

OPERATION = "gen_ai.operation.name"
"""Indicates the type of span: `call_llm`, `execute_tool`, `invoke_agent`."""

OUTPUT = "gen_ai.output"
"""Used in both LLM Calls and Tool Executions for holding their respective outputs."""

OUTPUT_TYPE = "gen_ai.output.type"
"""One of `json` or `text`. If `json`, the value of `OUTPUT` can be passed to `json.loads`."""

TOOL_ARGS = "gen_ai.tool.args"
"""Arguments passed to the executed tool."""

TOOL_DESCRIPTION = "gen_ai.tool.description"
"""Description of the executed tool."""

TOOL_NAME = "gen_ai.tool.name"
"""Name of the executed tool."""
