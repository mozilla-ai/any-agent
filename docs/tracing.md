# Agent Tracing

`any-agent` uses [`openinference`](https://github.com/Arize-ai/openinference) to generate
standardized [OpenTelemetry](https://opentelemetry.io/) traces for any of the supported [agent frameworks](./frameworks.md).

## Example

```py
from any_agent import AgentConfig, AgentFramework, AnyAgent
from any_agent.tracing import get_trace_provider, setup_tracing

framework = AgentFramework("openai")
agent = AnyAgent(
        main_agent=AgentConfig(
        model_id="gpt-4o",
        tools=["any_agent.tools.search_web", "any_agent.tools.visit_webpage"]
    )
)

trace_provider = get_trace_provider(project_name="example", agent_framework=framework)
setup_tracing(trace_provider, framework)
```
