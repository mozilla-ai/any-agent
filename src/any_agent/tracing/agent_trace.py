from collections.abc import Mapping
from datetime import timedelta
from functools import cached_property
from typing import TYPE_CHECKING, Any

from litellm.cost_calculator import cost_per_token
from pydantic import BaseModel, ConfigDict, Field

from any_agent.logging import logger

from .otel_types import (
    AttributeValue,
    Event,
    Link,
    Resource,
    SpanContext,
    SpanKind,
    Status,
)

if TYPE_CHECKING:
    from opentelemetry.sdk.trace import ReadableSpan


class TokenInfo(BaseModel):
    """Token Count information."""

    input_tokens: int
    output_tokens: int

    model_config = ConfigDict(extra="forbid")


class CostInfo(BaseModel):
    """Cost information."""

    input_cost: float
    output_cost: float

    model_config = ConfigDict(extra="forbid")


def compute_cost_info(
    attributes: Mapping[str, AttributeValue],
) -> CostInfo | None:
    """Use litellm to compute cost."""
    if not any(
        key in attributes
        for key in ["gen_ai.usage.input_tokens", "gen_ai.usage.output_tokens"]
    ):
        return None

    new_info: dict[str, float] = {}
    try:
        cost_prompt, cost_completion = cost_per_token(
            model=str(attributes.get("gen_ai.request.model", "")),
            prompt_tokens=int(attributes.get("gen_ai.usage.input_tokens", 0)),  # type: ignore[arg-type]
            completion_tokens=int(attributes.get("gen_ai.usage.output_tokens", 0)),  # type: ignore[arg-type]
        )
        new_info["input_cost"] = cost_prompt
        new_info["output_cost"] = cost_completion
    except Exception as e:
        msg = f"Error computing cost_per_token: {e}"
        logger.warning(msg)
        new_info["input_cost"] = 0.0
        new_info["output_cost"] = 0.0
    return CostInfo.model_validate(new_info)


class AgentSpan(BaseModel):
    """A span that can be exported to JSON or printed to the console."""

    name: str
    kind: SpanKind
    parent: SpanContext | None = None
    start_time: int | None = None
    end_time: int | None = None
    status: Status
    context: SpanContext
    attributes: dict[str, Any]
    links: list[Link]
    events: list[Event]
    resource: Resource

    model_config = ConfigDict(arbitrary_types_allowed=False)

    @classmethod
    def from_readable_span(cls, readable_span: "ReadableSpan") -> "AgentSpan":
        """Create an AgentSpan from a ReadableSpan."""
        return cls(
            name=readable_span.name,
            kind=SpanKind.from_otel(readable_span.kind),
            parent=SpanContext.from_otel(readable_span.parent),
            start_time=readable_span.start_time,
            end_time=readable_span.end_time,
            status=Status.from_otel(readable_span.status),
            context=SpanContext.from_otel(readable_span.context),
            attributes=dict(readable_span.attributes)
            if readable_span.attributes
            else {},
            links=[Link.from_otel(link) for link in readable_span.links],
            events=[Event.from_otel(event) for event in readable_span.events],
            resource=Resource.from_otel(readable_span.resource),
        )

    def add_cost_info(self) -> None:
        """Extend attributes with `TokenUseAndCost`."""
        cost_info = compute_cost_info(self.attributes)
        if cost_info:
            self.set_attributes(
                {f"gen_ai.usage.{k}": v for k, v in cost_info.model_dump().items()}
            )

    def set_attributes(self, attributes: Mapping[str, AttributeValue]) -> None:
        """Set attributes for the span."""
        for key, value in attributes.items():
            if key in self.attributes:
                logger.warning("Overwriting attribute %s with %s", key, value)
            self.attributes[key] = value

    def is_llm_call(self):
        """Check whether this span is a call to an LLM."""
        return self.attributes.get("gen_ai.operation.name") == "call_llm"


class AgentTrace(BaseModel):
    """A trace that can be exported to JSON or printed to the console."""

    spans: list[AgentSpan] = Field(default_factory=list)
    """A list of [`AgentSpan`][any_agent.tracing.trace.AgentSpan] that form the trace.
    """

    final_output: str | None = None
    """Contains the final output message returned by the agent.
    """

    model_config = ConfigDict(arbitrary_types_allowed=True)

    def _invalidate_usage_and_cost_cache(self) -> None:
        """Clear the cached usage_and_cost property if it exists."""
        if "usage" in self.__dict__:
            del self.usage
        if "cost" in self.__dict__:
            del self.cost

    def add_span(self, span: AgentSpan) -> None:
        """Add an AgentSpan to the trace and clear the usage_and_cost cache if present."""
        self.spans.append(span)
        self._invalidate_usage_and_cost_cache()

    def add_spans(self, spans: list[AgentSpan]) -> None:
        """Add a list of AgentSpans to the trace and clear the usage_and_cost cache if present."""
        self.spans.extend(spans)
        self._invalidate_usage_and_cost_cache()

    @property
    def duration(self) -> timedelta:
        """Returns the duration of the AGENT span named 'any_agent' as a datetime.timedelta object.

        The duration is computed from the span's start and end time (in nanoseconds).
        This functionality relies on the trace behavior where we create a span named 'any_agent' for each agent run,
        which is defined in each agent framework's `run` function.

        Raises ValueError if there are no spans, if the AGENT span is not found, or if start/end times are missing.
        """
        if not self.spans:
            msg = "No spans found in trace"
            raise ValueError(msg)
        for span in self.spans:
            if span.attributes.get("any_agent.run_id"):
                if span.start_time is not None and span.end_time is not None:
                    duration_ns = span.end_time - span.start_time
                    return timedelta(seconds=duration_ns / 1_000_000_000)
                msg = "Start or end time is missing for the AGENT span"
                raise ValueError(msg)
        msg = "Span with any_agent.run_id not found in trace"
        raise ValueError(msg)

    @cached_property
    def usage(self) -> TokenInfo:
        """The current total token usage statistics for this trace. Cached after first computation."""
        sum_input_tokens = 0
        sum_output_tokens = 0
        for span in self.spans:
            if span.is_llm_call():
                sum_input_tokens += span.attributes.get("gen_ai.usage.input_tokens", 0)
                sum_output_tokens += span.attributes.get(
                    "gen_ai.usage.output_tokens", 0
                )
        return TokenInfo(input_tokens=sum_input_tokens, output_tokens=sum_output_tokens)

    @cached_property
    def cost(self) -> CostInfo:
        """The current total cost statistics for this trace. Cached after first computation."""
        sum_input_cost = 0
        sum_output_cost = 0
        for span in self.spans:
            if span.is_llm_call():
                sum_input_cost += span.attributes.get("gen_ai.usage.input_cost", 0)
                sum_output_cost += span.attributes.get("gen_ai.usage.output_cost", 0)
        return CostInfo(input_cost=sum_input_cost, output_cost=sum_output_cost)
