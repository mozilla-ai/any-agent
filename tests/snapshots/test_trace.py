from any_agent.tracing.trace import AgentTrace
import json
from pathlib import Path

def test_openai_agent_trace_snapshot(snapshot):
    # Load the JSON file
    trace_path = Path(__file__).parent.parent / "assets" / "OPENAI.json"
    with open(trace_path, encoding="utf-8") as f:
        trace_data = json.load(f)
    # Try to parse as AgentTrace (will fail if schema changes)
    agent_trace = AgentTrace.model_validate(trace_data)
    # Snapshot the dict representation (so you see changes in the schema)
    # If this assert fails and you decide that you're ok with the new schema,
    # you can easily update the snapshot by running:
    # pytest tests/snapshots --snapshot-update
    assert agent_trace.model_dump() == snapshot
