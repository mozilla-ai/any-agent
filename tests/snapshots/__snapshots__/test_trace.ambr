# serializer version: 1
# name: test_openai_agent_trace_snapshot
  dict({
    'final_output': 'The current year in the America/New_York timezone is 2025. I have written this value to a file.',
    'spans': list([
      dict({
        'attributes': dict({
          'llm.system': 'openai',
          'openinference.span.kind': 'CHAIN',
          'output.mime_type': 'application/json',
          'output.value': '["get_current_time", "convert_time"]',
          'service.name': 'openai',
        }),
        'context': dict({
          'is_remote': False,
          'span_id': 13487556102999248805,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'end_time': 1747660964059570176,
        'events': list([
        ]),
        'kind': <SpanKind.INTERNAL: 'internal'>,
        'links': list([
        ]),
        'name': 'mcp_tools',
        'parent': dict({
          'is_remote': False,
          'span_id': 1351128178870435741,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'resource': dict({
          'attributes': dict({
            'service.name': 'unknown_service',
            'telemetry.sdk.language': 'python',
            'telemetry.sdk.name': 'opentelemetry',
            'telemetry.sdk.version': '1.33.1',
          }),
          'schema_url': '',
        }),
        'start_time': 1747660964058813952,
        'status': dict({
          'description': None,
          'status_code': <StatusCode.OK: 'ok'>,
        }),
      }),
      dict({
        'attributes': dict({
          'cost_completion': 0.0,
          'cost_prompt': 0.00011439999999999999,
          'input.mime_type': 'application/json',
          'input.value': '[{"content": "Search the web to answer", "role": "system"}, {"role": "user", "content": "Use the tools to find what year it is in the America/New_York timezone and write the value (single number) to a file"}]',
          'llm.input_messages.0.message.content': 'Search the web to answer',
          'llm.input_messages.0.message.role': 'system',
          'llm.input_messages.1.message.content': 'Use the tools to find what year it is in the America/New_York timezone and write the value (single number) to a file',
          'llm.input_messages.1.message.role': 'user',
          'llm.invocation_parameters': '{"temperature": 0.0, "parallel_tool_calls": false, "base_url": "", "model_impl": "litellm"}',
          'llm.model_name': 'gpt-4.1-mini',
          'llm.output_messages.0.message.role': 'assistant',
          'llm.output_messages.0.message.tool_calls.0.tool_call.function.arguments': '{"timezone":"America/New_York"}',
          'llm.output_messages.0.message.tool_calls.0.tool_call.function.name': 'get_current_time',
          'llm.output_messages.0.message.tool_calls.0.tool_call.id': 'call_7uZkRepA70GgAS5sT5mCHfNj',
          'llm.system': 'openai',
          'llm.token_count.completion': 18,
          'llm.token_count.prompt': 286,
          'openinference.span.kind': 'LLM',
          'output.mime_type': 'application/json',
          'output.value': '[{"content": null, "role": "assistant", "tool_calls": [{"function": {"arguments": "{\\"timezone\\":\\"America/New_York\\"}", "name": "get_current_time"}, "id": "call_7uZkRepA70GgAS5sT5mCHfNj", "type": "function"}], "function_call": null, "annotations": []}]',
          'service.name': 'openai',
        }),
        'context': dict({
          'is_remote': False,
          'span_id': 4343523433967492499,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'end_time': 1747660968844551936,
        'events': list([
        ]),
        'kind': <SpanKind.INTERNAL: 'internal'>,
        'links': list([
        ]),
        'name': 'generation',
        'parent': dict({
          'is_remote': False,
          'span_id': 1351128178870435741,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'resource': dict({
          'attributes': dict({
            'service.name': 'unknown_service',
            'telemetry.sdk.language': 'python',
            'telemetry.sdk.name': 'opentelemetry',
            'telemetry.sdk.version': '1.33.1',
          }),
          'schema_url': '',
        }),
        'start_time': 1747660964060802816,
        'status': dict({
          'description': None,
          'status_code': <StatusCode.OK: 'ok'>,
        }),
      }),
      dict({
        'attributes': dict({
          'input.mime_type': 'application/json',
          'input.value': '{"timezone":"America/New_York"}',
          'llm.system': 'openai',
          'openinference.span.kind': 'TOOL',
          'output.mime_type': 'application/json',
          'output.value': '{"type":"text","text":"{\\n  \\"timezone\\": \\"America/New_York\\",\\n  \\"datetime\\": \\"2025-05-19T09:22:48-04:00\\",\\n  \\"is_dst\\": true\\n}","annotations":null}',
          'service.name': 'openai',
          'tool.name': 'get_current_time',
        }),
        'context': dict({
          'is_remote': False,
          'span_id': 6439485983011741764,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'end_time': 1747660968863655168,
        'events': list([
        ]),
        'kind': <SpanKind.INTERNAL: 'internal'>,
        'links': list([
        ]),
        'name': 'get_current_time',
        'parent': dict({
          'is_remote': False,
          'span_id': 1351128178870435741,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'resource': dict({
          'attributes': dict({
            'service.name': 'unknown_service',
            'telemetry.sdk.language': 'python',
            'telemetry.sdk.name': 'opentelemetry',
            'telemetry.sdk.version': '1.33.1',
          }),
          'schema_url': '',
        }),
        'start_time': 1747660968860654080,
        'status': dict({
          'description': None,
          'status_code': <StatusCode.OK: 'ok'>,
        }),
      }),
      dict({
        'attributes': dict({
          'cost_completion': 0.0,
          'cost_prompt': 0.00014759999999999998,
          'input.mime_type': 'application/json',
          'input.value': '[{"content": "Search the web to answer", "role": "system"}, {"role": "user", "content": "Use the tools to find what year it is in the America/New_York timezone and write the value (single number) to a file"}, {"role": "assistant", "tool_calls": [{"id": "call_7uZkRepA70GgAS5sT5mCHfNj", "type": "function", "function": {"name": "get_current_time", "arguments": "{\\"timezone\\":\\"America/New_York\\"}"}}]}, {"role": "tool", "tool_call_id": "call_7uZkRepA70GgAS5sT5mCHfNj", "content": "{\\"type\\":\\"text\\",\\"text\\":\\"{\\\\n  \\\\\\"timezone\\\\\\": \\\\\\"America/New_York\\\\\\",\\\\n  \\\\\\"datetime\\\\\\": \\\\\\"2025-05-19T09:22:48-04:00\\\\\\",\\\\n  \\\\\\"is_dst\\\\\\": true\\\\n}\\",\\"annotations\\":null}"}]',
          'llm.input_messages.0.message.content': 'Search the web to answer',
          'llm.input_messages.0.message.role': 'system',
          'llm.input_messages.1.message.content': 'Use the tools to find what year it is in the America/New_York timezone and write the value (single number) to a file',
          'llm.input_messages.1.message.role': 'user',
          'llm.input_messages.2.message.role': 'assistant',
          'llm.input_messages.2.message.tool_calls.0.tool_call.function.arguments': '{"timezone":"America/New_York"}',
          'llm.input_messages.2.message.tool_calls.0.tool_call.function.name': 'get_current_time',
          'llm.input_messages.2.message.tool_calls.0.tool_call.id': 'call_7uZkRepA70GgAS5sT5mCHfNj',
          'llm.input_messages.3.message.content': '{"type":"text","text":"{\\n  \\"timezone\\": \\"America/New_York\\",\\n  \\"datetime\\": \\"2025-05-19T09:22:48-04:00\\",\\n  \\"is_dst\\": true\\n}","annotations":null}',
          'llm.input_messages.3.message.role': 'tool',
          'llm.input_messages.3.message.tool_call_id': 'call_7uZkRepA70GgAS5sT5mCHfNj',
          'llm.invocation_parameters': '{"temperature": 0.0, "parallel_tool_calls": false, "base_url": "", "model_impl": "litellm"}',
          'llm.model_name': 'gpt-4.1-mini',
          'llm.output_messages.0.message.role': 'assistant',
          'llm.output_messages.0.message.tool_calls.0.tool_call.function.arguments': '{"text":"2025"}',
          'llm.output_messages.0.message.tool_calls.0.tool_call.function.name': 'write_file',
          'llm.output_messages.0.message.tool_calls.0.tool_call.id': 'call_v6mRm3m24xfm3dftlRbkgoce',
          'llm.system': 'openai',
          'llm.token_count.completion': 15,
          'llm.token_count.prompt': 369,
          'openinference.span.kind': 'LLM',
          'output.mime_type': 'application/json',
          'output.value': '[{"content": null, "role": "assistant", "tool_calls": [{"function": {"arguments": "{\\"text\\":\\"2025\\"}", "name": "write_file"}, "id": "call_v6mRm3m24xfm3dftlRbkgoce", "type": "function"}], "function_call": null, "annotations": []}]',
          'service.name': 'openai',
        }),
        'context': dict({
          'is_remote': False,
          'span_id': 17098691010651300468,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'end_time': 1747660969739481856,
        'events': list([
        ]),
        'kind': <SpanKind.INTERNAL: 'internal'>,
        'links': list([
        ]),
        'name': 'generation',
        'parent': dict({
          'is_remote': False,
          'span_id': 1351128178870435741,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'resource': dict({
          'attributes': dict({
            'service.name': 'unknown_service',
            'telemetry.sdk.language': 'python',
            'telemetry.sdk.name': 'opentelemetry',
            'telemetry.sdk.version': '1.33.1',
          }),
          'schema_url': '',
        }),
        'start_time': 1747660968868616960,
        'status': dict({
          'description': None,
          'status_code': <StatusCode.OK: 'ok'>,
        }),
      }),
      dict({
        'attributes': dict({
          'input.mime_type': 'application/json',
          'input.value': '{"text":"2025"}',
          'llm.system': 'openai',
          'openinference.span.kind': 'TOOL',
          'service.name': 'openai',
          'tool.name': 'write_file',
        }),
        'context': dict({
          'is_remote': False,
          'span_id': 12723702095548437253,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'end_time': 1747660969746458112,
        'events': list([
        ]),
        'kind': <SpanKind.INTERNAL: 'internal'>,
        'links': list([
        ]),
        'name': 'write_file',
        'parent': dict({
          'is_remote': False,
          'span_id': 1351128178870435741,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'resource': dict({
          'attributes': dict({
            'service.name': 'unknown_service',
            'telemetry.sdk.language': 'python',
            'telemetry.sdk.name': 'opentelemetry',
            'telemetry.sdk.version': '1.33.1',
          }),
          'schema_url': '',
        }),
        'start_time': 1747660969745337088,
        'status': dict({
          'description': None,
          'status_code': <StatusCode.OK: 'ok'>,
        }),
      }),
      dict({
        'attributes': dict({
          'cost_completion': 0.0,
          'cost_prompt': 0.0001572,
          'input.mime_type': 'application/json',
          'input.value': '[{"content": "Search the web to answer", "role": "system"}, {"role": "user", "content": "Use the tools to find what year it is in the America/New_York timezone and write the value (single number) to a file"}, {"role": "assistant", "tool_calls": [{"id": "call_7uZkRepA70GgAS5sT5mCHfNj", "type": "function", "function": {"name": "get_current_time", "arguments": "{\\"timezone\\":\\"America/New_York\\"}"}}]}, {"role": "tool", "tool_call_id": "call_7uZkRepA70GgAS5sT5mCHfNj", "content": "{\\"type\\":\\"text\\",\\"text\\":\\"{\\\\n  \\\\\\"timezone\\\\\\": \\\\\\"America/New_York\\\\\\",\\\\n  \\\\\\"datetime\\\\\\": \\\\\\"2025-05-19T09:22:48-04:00\\\\\\",\\\\n  \\\\\\"is_dst\\\\\\": true\\\\n}\\",\\"annotations\\":null}"}, {"role": "assistant", "tool_calls": [{"id": "call_v6mRm3m24xfm3dftlRbkgoce", "type": "function", "function": {"name": "write_file", "arguments": "{\\"text\\":\\"2025\\"}"}}]}, {"role": "tool", "tool_call_id": "call_v6mRm3m24xfm3dftlRbkgoce", "content": "None"}]',
          'llm.input_messages.0.message.content': 'Search the web to answer',
          'llm.input_messages.0.message.role': 'system',
          'llm.input_messages.1.message.content': 'Use the tools to find what year it is in the America/New_York timezone and write the value (single number) to a file',
          'llm.input_messages.1.message.role': 'user',
          'llm.input_messages.2.message.role': 'assistant',
          'llm.input_messages.2.message.tool_calls.0.tool_call.function.arguments': '{"timezone":"America/New_York"}',
          'llm.input_messages.2.message.tool_calls.0.tool_call.function.name': 'get_current_time',
          'llm.input_messages.2.message.tool_calls.0.tool_call.id': 'call_7uZkRepA70GgAS5sT5mCHfNj',
          'llm.input_messages.3.message.content': '{"type":"text","text":"{\\n  \\"timezone\\": \\"America/New_York\\",\\n  \\"datetime\\": \\"2025-05-19T09:22:48-04:00\\",\\n  \\"is_dst\\": true\\n}","annotations":null}',
          'llm.input_messages.3.message.role': 'tool',
          'llm.input_messages.3.message.tool_call_id': 'call_7uZkRepA70GgAS5sT5mCHfNj',
          'llm.input_messages.4.message.role': 'assistant',
          'llm.input_messages.4.message.tool_calls.1.tool_call.function.arguments': '{"text":"2025"}',
          'llm.input_messages.4.message.tool_calls.1.tool_call.function.name': 'write_file',
          'llm.input_messages.4.message.tool_calls.1.tool_call.id': 'call_v6mRm3m24xfm3dftlRbkgoce',
          'llm.input_messages.5.message.content': 'None',
          'llm.input_messages.5.message.role': 'tool',
          'llm.input_messages.5.message.tool_call_id': 'call_v6mRm3m24xfm3dftlRbkgoce',
          'llm.invocation_parameters': '{"temperature": 0.0, "parallel_tool_calls": false, "base_url": "", "model_impl": "litellm"}',
          'llm.model_name': 'gpt-4.1-mini',
          'llm.output_messages.0.message.content': 'The current year in the America/New_York timezone is 2025. I have written this value to a file.',
          'llm.output_messages.0.message.role': 'assistant',
          'llm.system': 'openai',
          'llm.token_count.completion': 25,
          'llm.token_count.prompt': 393,
          'openinference.span.kind': 'LLM',
          'output.mime_type': 'application/json',
          'output.value': '[{"content": "The current year in the America/New_York timezone is 2025. I have written this value to a file.", "role": "assistant", "tool_calls": null, "function_call": null, "annotations": []}]',
          'service.name': 'openai',
        }),
        'context': dict({
          'is_remote': False,
          'span_id': 8789522812412071271,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'end_time': 1747660970767698176,
        'events': list([
        ]),
        'kind': <SpanKind.INTERNAL: 'internal'>,
        'links': list([
        ]),
        'name': 'generation',
        'parent': dict({
          'is_remote': False,
          'span_id': 1351128178870435741,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'resource': dict({
          'attributes': dict({
            'service.name': 'unknown_service',
            'telemetry.sdk.language': 'python',
            'telemetry.sdk.name': 'opentelemetry',
            'telemetry.sdk.version': '1.33.1',
          }),
          'schema_url': '',
        }),
        'start_time': 1747660969748457984,
        'status': dict({
          'description': None,
          'status_code': <StatusCode.OK: 'ok'>,
        }),
      }),
      dict({
        'attributes': dict({
          'llm.system': 'openai',
          'openinference.span.kind': 'AGENT',
          'service.name': 'openai',
        }),
        'context': dict({
          'is_remote': False,
          'span_id': 1351128178870435741,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'end_time': 1747660970773938944,
        'events': list([
        ]),
        'kind': <SpanKind.INTERNAL: 'internal'>,
        'links': list([
        ]),
        'name': 'any_agent',
        'parent': dict({
          'is_remote': False,
          'span_id': 701147585678663371,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'resource': dict({
          'attributes': dict({
            'service.name': 'unknown_service',
            'telemetry.sdk.language': 'python',
            'telemetry.sdk.name': 'opentelemetry',
            'telemetry.sdk.version': '1.33.1',
          }),
          'schema_url': '',
        }),
        'start_time': 1747660964058574080,
        'status': dict({
          'description': None,
          'status_code': <StatusCode.OK: 'ok'>,
        }),
      }),
      dict({
        'attributes': dict({
          'openinference.span.kind': 'AGENT',
          'service.name': 'openai',
        }),
        'context': dict({
          'is_remote': False,
          'span_id': 701147585678663371,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'end_time': 1747660970774269000,
        'events': list([
        ]),
        'kind': <SpanKind.INTERNAL: 'internal'>,
        'links': list([
        ]),
        'name': 'Agent workflow',
        'parent': dict({
          'is_remote': False,
          'span_id': 1496043382881042842,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'resource': dict({
          'attributes': dict({
            'service.name': 'unknown_service',
            'telemetry.sdk.language': 'python',
            'telemetry.sdk.name': 'opentelemetry',
            'telemetry.sdk.version': '1.33.1',
          }),
          'schema_url': '',
        }),
        'start_time': 1747660964058293000,
        'status': dict({
          'description': None,
          'status_code': <StatusCode.OK: 'ok'>,
        }),
      }),
      dict({
        'attributes': dict({
          'any_agent.run_id': 'e52f73a4-0d70-4182-b5a7-c60bff6ff428',
          'service.name': 'openai',
        }),
        'context': dict({
          'is_remote': False,
          'span_id': 1496043382881042842,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': 312240062924141420731563990691240884516,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'end_time': 1747660970774416000,
        'events': list([
        ]),
        'kind': <SpanKind.INTERNAL: 'internal'>,
        'links': list([
        ]),
        'name': 'agent_run',
        'parent': dict({
          'is_remote': False,
          'span_id': None,
          'trace_flags': dict({
            'value': 0,
          }),
          'trace_id': None,
          'trace_state': dict({
            'entries': dict({
            }),
          }),
        }),
        'resource': dict({
          'attributes': dict({
            'service.name': 'unknown_service',
            'telemetry.sdk.language': 'python',
            'telemetry.sdk.name': 'opentelemetry',
            'telemetry.sdk.version': '1.33.1',
          }),
          'schema_url': '',
        }),
        'start_time': 1747660964057285000,
        'status': dict({
          'description': None,
          'status_code': <StatusCode.UNSET: 'unset'>,
        }),
      }),
    ]),
  })
# ---
